core/app/models/spree/order_contents.rb

```diff
@@ -57,6 +71,20 @@ def approve(user: nil, name: nil)
 
     private
 
+    def should_recalculate_taxes?(old_address, new_address)
+      # Possible cases:
+      # [Old_address is a TaxLocation, New_address is a TaxLocation ]:
+      # NON-VAT order before Address state : we don't compute taxes.
+      # [Old_address is a TaxLocation, New_address is an Adress ]:
+      # Order who passed Address state : we compute taxes.
+      # [Old_address is an Adress, New_address is a TaxLocation ]:
+      # Should not happen, TaxLocation can be replaced by an Address but not the other way around.
+      # Order is a weird state and we do nothing because it should not happen.
+      return if old_address.is_a?(Spree::Tax::TaxLocation) || new_address.is_a?(Spree::Tax::TaxLocation)
+
+      old_address.try(:taxation_attributes) != new_address.try(:taxation_attributes)
```

Comment:
Sorry for the delay in reviewing. Could these be `.try!`?

Otherwise I think this is a great change and is a step in the right direction towards relying less on the state machine to get everything right.